hail/src/main/scala/is/hail/expr/ir/TypeCheck.scala

```diff
@@ -341,9 +341,16 @@ object TypeCheck {
         assert(coerce[TStream](x.typ).elementType == join.typ)
         assert(lKey.forall(lEltTyp.hasField))
         assert(rKey.forall(rEltTyp.hasField))
-        assert((lKey, rKey).zipped.forall { case (lk, rk) =>
-          lEltTyp.fieldType(lk) == rEltTyp.fieldType(rk)
-        })
+        if (x.isIntervalJoin) {
+          val lKeyTyp = lEltTyp.fieldType(lKey(0))
+          val rKeyTyp = rEltTyp.fieldType(rKey(0)).asInstanceOf[TInterval]
+          assert(lKeyTyp == rKeyTyp.pointType)
+          assert((joinType == "left") || (joinType == "inner"))
```

Comment:
Since these asserts are here, shouldn't the `isIntervalJoin` function on `StreamJoinRightDistinct` have being `"left" or "inner" as conditions of it being true?

Reply:
I agree it's a little weird. My thought process was that `isIntervalJoin` assumes a well-typed node. If the join type isn't "left" or "inner", but the right side is interval keyed while the left isn't, the right answer isn't "isIntervalJoin = false", it's just a malformed IR. I think that's the right general philosophy: outside of TypeCheck, assume things are well-typed, so we don't have to litter the same typechecking assertions everywhere.